csharp/src/Ice/Logger.cs

```diff
@@ -8,161 +8,123 @@
-            _file = file;
-            _writer = new StreamWriter(new FileStream(file, FileMode.Append, FileAccess.Write, FileShare.ReadWrite));
-        }
+        public virtual void Trace(string category, string message) => Write(Format("--", category, message));
 
-        public override ILogger CloneWithPrefix(string prefix) => new FileLogger(prefix, _file);
+        public virtual void Warning(string message) => Write(Format("-!", "warning", message));
 
-        protected override void Write(string message)
+        /// <summary>Creates a new logger, used only by derived classes.</summary>
+        /// <param name="prefix">The prefix to perpend to messages write by this logger.</param>
+        protected Logger(string prefix)
         {
-            _writer.WriteLine(message);
-            _writer.Flush();
+            Prefix = prefix;
+            FormattedPrefix = prefix.Length > 0 ? $"{prefix}: " : "";
         }
 
-        public void Destroy() => _writer.Close();
-
-        private readonly string _file;
-        private readonly TextWriter _writer;
-    }
-
-    public class ConsoleListener : TraceListener
-    {
-        public override bool IsThreadSafe => true;
+        /// <summary>Writes a message using this logger.</summary>
+        /// <param name="message">The message to write.</param>
+        protected abstract void Write(string message);
 
-        public override void TraceEvent(TraceEventCache cache, string source, TraceEventType type,
-                                        int id, string message)
+        private string Format(string prefix, string category, string message)
         {
-            System.Text.StringBuilder s;
-            if (type == TraceEventType.Error)
-            {
-                s = new System.Text.StringBuilder("!!");
-            }
-            else if (type == TraceEventType.Warning)
-            {
-                s = new System.Text.StringBuilder("-!");
-            }
-            else
-            {
-                s = new System.Text.StringBuilder("--");
-            }
+            var s = new System.Text.StringBuilder(prefix);
             s.Append(' ');
             s.Append(System.DateTime.Now.ToString(Date, CultureInfo.CurrentCulture));
             s.Append(' ');
             s.Append(System.DateTime.Now.ToString(Time, CultureInfo.CurrentCulture));
             s.Append(' ');
+            s.Append(FormattedPrefix);
+            s.Append(category);
+            s.Append(": ");
             s.Append(message);
-            WriteLine(s.ToString());
+            s.Replace("\n", "\n   ");
+            return s.ToString();
         }
-
-        public override void Write(string message) => System.Console.Error.Write(message);
-
-        public override void WriteLine(string message) => System.Console.Error.WriteLine(message);
-
-        private const string Date = "d";
-        private const string Time = "HH:mm:ss:fff";
     }
 
     public sealed class TraceLogger : Logger
```

Comment:
What's the purpose of a TraceLogger?

Reply:
This logger publish log messages using the Trace api, .NET send the message to all the listeners configured, we have a console listener that writes to the console, but users can add other listeners programmatically or with .NET configuration properties

https://docs.microsoft.com/en-us/dotnet/framework/debug-trace-profile/trace-listeners

We have  a property that controls when the console listener is added

https://doc.zeroc.com/ice/3.7/property-reference/miscellaneous-ice-properties#id-.MiscellaneousIce.*Propertiesv3.7-Ice.ConsoleListener

Reply:
The property documentation is a bit confusing, ` If num is 0, logging is disabled.`, this doesn't disable logging, when set to 0 the `ConsoleListener` that writes the messages to the standard error is not added to the trace listeners, but logger is still enabled and will be send to any listeners added by the application